### PR TITLE
feat(provider-generator): Use `ComplexList` for any complex list

### LINKED ISSUE #25: Complex Computed Optional Types
There are optional computed complex types in the provider schema, e.g. in the [aws_security_group](https://www.terraform.io/docs/providers/aws/r/security_group.html) the attributes `egress` / `ingress`. Here an excerpt from the schema:

```
             "egress": {
                "type": [
                  "set",
                  [
                    "object",
                    {
                      // ... attributes 
                    }
                  ]
                ],
                "optional": true,
                "computed": true
              },
```

As I understand `optional` && `computed`, this means there could be `egress` or `ingress` rules created by AWS. Which, at least for `egress` is true. However, apparently the AWS Terraform provider removes this `computed` object (see [comment](https://www.terraform.io/docs/providers/aws/r/security_group.html#description-2))

At the moment the attribute is rendered without taking into account the `computed` option, which looks like this:

```typescript
  // egress
  private _egress?: SecurityGroupEgress[];
  public get egress() {
    return this._egress; // Getting the computed value is not yet implemented
  }
  public set egress(value: SecurityGroupEgress[]) {
    this._egress = value;
  }
```

We should probably enable access to the those `computed` rules as well. Something along the lines of this:

```typescript
  // egress
  private _egress?: SecurityGroupEgress[];
  public get egress(index?: string) {
    return this._egress || return new SecurityGroupEgressList(this, 'egress', index);
  }
  public set egress(value: SecurityGroupEgress[] | undefined) {
    this._egress = value;
  }
```

We'd have to generate:

- an interface (e.g. `SecurityGroupEgress`) for the `optional` type 
- a class (e.g. `SecurityGroupEgressList`) which implements the interface to wrap the `computed` value
- Make sure the return type of the getter function is intuitively usable

In particular the last point seems to be difficult.

For now it's probably fine as is, but when heading towards beta we should come back to this.

### LINKED ISSUE #434: ContainerPorts in Docker Provider Container getter returns string identifier rather than container ports
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### cdktf & Language Versions
Terraform v0.13.2
0.0.18

### Affected Resource(s)

Docker provider Container

### Expected Behavior

Container.ports should return a list of ContainerPorts.

### Actual Behavior

Container.ports returns the fqn of the ContainerPorts object. 
console.log(container.ports);
// Log output is ${docker_container.container-name.ports}

### Steps to Reproduce

Create a container resource with some ports, then try to access the ports later.

### Important Factoids

I ran in to this trying to write an aspect to verify that all containers had unique external ports on their network. I think that the provider code is maybe incorrect, it returns interpolationForAttribute('ports') as any. It could return the private backing field _ports, or the attribute lookup could be updated. I tried retuning getListAttribute, but I got a list of Token (i.e. #{Token[TOKEN.1]}).

### LINKED ISSUE #518: AzureRM CosmosDB connection_strings[0] yields #{Token[TOKEN.0]} instead of first list item
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### cdktf & Language Versions

`cdktf`: `0.1.0`
`@cdktf/provider-azurerm`: `0.0.91`

### Affected Resource(s)

`CosmosdbAccount` from `@cdktf/provider-azurerm`

### Expected Behavior

Referencing `cosmosDbAccount.connectionStrings[0]`, e.g. as a terraform output, should yield the actual reference (i.e. in HCL: `"${azurerm_cosmosdb_account.cosmosdb_account.connection_strings[0]}"`).

### Actual Behavior

It actually yields `"#{Token[TOKEN.0]}"`

### Important Factoids

I'm pretty sure that this is closely related to #434 (and thus probably also #435), but I didn't want to hijack any of those for my question. I'm not very familiar with CDKTF yet, but from what I've read, I was expecting that `getListAttribute` was implemented for that exact purpose, but it doesn't seem to work here. Is this a bug or did I just miss something? And if not, is there a workaround for that?

### References

* #434

### LINKED ISSUE #663: EKS can't get OIDC issuer
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### cdktf & Language Versions

```
{
  "language": "typescript",
  "app": "npm run --silent compile && node main.js",
  "terraformProviders": [
    "aws@~> 2.0",
    "azurerm@~> 2.0.0"
  ]
}
```

- cdktf version
```
cdktf --version
0.3.0
```

### Affected Resource(s)

EksCluster

### Expected Behavior

Get EKS oidc url

### Actual Behavior

Want to get oidc issuer from EKS

```
new IamOpenidConnectProvider(this, 'EksOidc', {
    clientIdList: ["sts.amazonaws.com"],
    thumbprintList: [],
    url: eks.identity('0').oidc('0').issuer
})
```

But `EksClusterIdentity` class not include `issuer` function.

```
    /cdktf-samples/lib/aws/kubernetes.js:141
                url: eks.identity('0').oidc('0').issuer
                                       ^

    TypeError: eks.identity(...).oidc is not a function
        at new AwsEksGroups (/cdktf-samples/lib/aws/kubernetes.js:141:36)
        at new AwsEks (/cdktf-samples/main.js:81:9)
        at Object.<anonymous> (/cdktf-samples/main.js:121:1)
        at Module._compile (internal/modules/cjs/loader.js:1063:30)
        at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
        at Module.load (internal/modules/cjs/loader.js:928:32)
        at Function.Module._load (internal/modules/cjs/loader.js:769:14)
        at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:72:12)
        at internal/main/run_main_module.js:17:47
```

In fact, `EksClusterIdentityOidc` included `issuer` but not any resource use it.

### Steps to Reproduce

cdktf deploy

### Workaround

new `EksClusterIdentityOidc` class with EKS resource

```
new IamOpenidConnectProvider(this, 'EksOidc', {
    clientIdList: ["sts.amazonaws.com"],
    thumbprintList: [],
    url: new EksClusterIdentityOidc(eks, 'identity[0]', 'oidc[0]').issuer
})
```

### References

none

### LINKED ISSUE #1087: Get one element from module output list(target_group_arns of terraform-aws-modules/alb/aws)
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### cdktf & Language Versions
cdktf 0.6.3
cdktf-cdktf-provider-aws 2.0.11
terraform-aws-modules/alb/aws@5.16.0
Python 3.8.10

### Affected Resource(s)

- aws alb module(TerraformAwsModulesAlbAws - https://registry.terraform.io/modules/terraform-aws-modules/alb/aws/latest)
- EcsService

I created several aws albs via module and want to get single element from their output list (target_group_arns) to create subsequent resources(ecs services).
I tried different approaches, the last one was as following(for test purposes I hardcoded index values):
```
lb = self.load_balancers[0]
target_group_arn = Fn.element(Token.as_list(lb.target_group_arns_output), 0)  
```
### Expected Behavior
I thought, that in json output will appear something like:
`"target_group_arn": "${element(module.stack-name-lb2.target_group_arns, 0)}"`
And then this string will be substituted with needed value during apply

### Actual Behavior

I can synthesize code successfully, but during the plan I get several errors like this one:

> Error: Invalid character
> │ 
> │   on cdk.tf.json line 2704, in resource.aws_ecs_service.stack-nameecs-cluster-construct-service-33_EEFE3F15.load_balancer[0]:
> │ 2704:             "target_group_arn": "${element(${module.stack-name-lb2.target_group_arns}, 0)}"
> │ 
> │ This character is not used within the language.
> ╵
> ╷
> │ Error: Invalid expression
> │ 
> │   on cdk.tf.json line 2704, in resource.aws_ecs_service.stack-nameecs-cluster-construct-service-33_EEFE3F15.load_balancer[0]:
> │ 2704:             "target_group_arn": "${element(${module.stack-name-lb2.target_group_arns}, 0)}"
> │ 
> │ Expected the start of an expression, but found an invalid expression token.

### LINKED ISSUE #1363: Go: Referencing boolean attributes in boolean arguments does not work
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### cdktf & Language Versions

* 0.7

### Affected Resource(s)

any

### Debug Output

See https://github.com/ansgarm/test-cdktf-go-bool-refs

### Expected Behavior

Should synthesize to a valid reference (works in e.g. TypeScript)

### Actual Behavior

Fails with 
```
panic: unable to make an instance of unregistered interface interface {}
```
(full error in reproduction example)

### Steps to Reproduce
https://github.com/ansgarm/test-cdktf-go-bool-refs

### Important Factoids

* also fails in Java, works in TS, Python & C#

### LINKED ISSUE #1371: Go: Referencing list attributes does not work
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### cdktf & Language Versions

* 0.7

### Affected Resource(s)

any

### Debug Output

https://github.com/ansgarm/test-cdktf-go-list-refs-k8s

### Expected Behavior

Should synthesize to a valid reference (works in e.g. TypeScript)

### Actual Behavior

Fails with 
```
panic: unable to make an instance of unregistered interface interface {}
```
(full error in reproduction example)

### Steps to Reproduce
https://github.com/ansgarm/test-cdktf-go-list-refs-k8s

### LINKED ISSUE #1447: Using the same reference in different locations (in and outside a TF Expression) causes issues
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### cdktf & Language Versions

- cdktf 0.8
- all languages

### Expected Behavior

When I use the same reference twice but in different settings (in or outside a TF Expression) the reference gets rendered in the same way (depending on the order of the references).

```ts

const myS3Bucket = new S3Bucket(this, "mybucket", {});
const arn = myS3Bucket.arn;


const a = Fn.join(",", [arn, `this is the arn: ${arn}`]) 
const b = Fn.join(",", [`this is the arn: ${arn}`, arn]) 
```

both a and b should be valid terraform but they are not

### Actual Behavior

The order of the arns leads either to one `${}` part outside of a string or a missing `${}` block inside the synthesised json.

### LINKED ISSUE #1520: token usage dependent on the order they are used (new with v0.9.0)
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### cdktf & Language Versions

python
v0.9.0 (bug not present in v0.8.6)

### Affected Resource(s)

all?

### Expected Behavior

token usage should be independent on the order they are used

### Actual Behavior

when tokens are nested then used out of order, they dont resolve properly

### Steps to Reproduce

```py
from constructs import Construct
from cdktf import App, TerraformStack, TerraformVariable, Fn

from imports.google import (
    BigqueryDataset,
    GoogleProvider,
)


class MyStack(TerraformStack):
    def __init__(self, scope: Construct, ns: str):
        super().__init__(scope, ns)
        google = GoogleProvider(self, "google")

        some_replacement = TerraformVariable(self, "some_replacement", type="string").string_value

        my_token = Fn.replace("bar REPLACE_ME", "REPLACE_ME", some_replacement)

        my_other_token = Fn.jsonencode(dict(foo=my_token))

        ds = BigqueryDataset(
            self,
            "test99_bq_dataset",
            provider=google,
            description=my_other_token,
            project="my-project",
            dataset_id="test99",
            location="EU",
        )

        ds2 = BigqueryDataset(
            self,
            "test9_bq_dataset",
            provider=google,
            description=my_token,
            project="my-project",
            dataset_id="test99",
            location="EU",
        )


def main():
    app = App()
    MyStack(app, "my_stack")
    app.synth()


if __name__ == "__main__":
    main()

```

above code synthesises:
```json
{
  ...
  "resource": {
    "google_bigquery_dataset": {
      "test99_bq_dataset": {
        "dataset_id": "test99",
        "description": "${jsonencode({foo = replace(\"bar REPLACE_ME\", \"REPLACE_ME\", var.some_replacement)})}",
        ...
      },
      "test9_bq_dataset": {
        "dataset_id": "test99",
        "description": "replace(\"bar REPLACE_ME\", \"REPLACE_ME\", var.some_replacement)",
        ...
      }
    }
  }
}
```

note that the description for "test9_bq_dataset" will not resolve because of the missing `${...}`

### Important Factoids

the reason for this (which took me ages to figure out!) is because `my_token` has previously been used in the definition for `my_other_token`

moving `ds2 = BigqueryDataset(...)` above `ds = BigqueryDataset(...)` in main.py synthesises the correct result

### References

- unexpected breaking change as a result of https://github.com/hashicorp/terraform-cdk/releases/tag/v0.9.0

### LINKED ISSUE #1610: Use `ComplexList` classes for references to list attributes of resources
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

[#1499 Model complex lists as arrays](https://github.com/hashicorp/terraform-cdk/pull/1499) introduces the `ComplexList` class which allows users to call `.get(index)` on computed lists (computed = you can't pass an input value for this attribute). This allows users to write e.g.
```ts
cert.domainValidationOptions.get(0).resourceRecordName
```

By using the same mechanism for the getter values of other list attributes, we can use the same `.get(index).someAttribute` instead of requiring the use of Terraform functions. See the following example for how complicated that currently is:
https://github.com/hashicorp/terraform-cdk/blob/d48ab89125165ed1c7079773c69a5f7fd7ac3118/test/typescript/edge/main.ts#L55-L62

This would require asymmetric getters and setters which is not directly supported by TypeScript, but can be worked around by replacing the setter with a `putSomeAttribute` method. We currently already do this for lists with single items:
https://github.com/hashicorp/terraform-cdk/blob/d48ab89125165ed1c7079773c69a5f7fd7ac3118/test/typescript/providers/main.ts#L186-L187

There might be other downsides to this proposal, so please feel free to comment and link other related issues.

### References

- https://github.com/hashicorp/terraform-cdk/pull/1499
- https://github.com/hashicorp/terraform-cdk/issues/25 (could actually be a duplicate, but I wasn't sure)

### LINKED ISSUE #1684: Nested tokens are not escaped when `Fn.format` is used on them
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### cdktf & Language Versions

* NodeJS 16
* cdktf 0.10.1

### Affected Resource(s)

Probably any, but I just tested with `StringResource` from `@cdktf/provider-random`

### Debug Output

<details><summary>debug output</summary>

```
yarn run v1.22.18
$ CDKTF_LOG_LEVEL=debug cdktf synth
[2022-04-04T17:41:21.284] [DEBUG] default - {
  "terraform_version": "1.1.7",
  "platform": "darwin_amd64",
  "provider_selections": {},
  "terraform_outdated": false
}

[2022-04-04T17:41:21.320] [DEBUG] default - v17.8.0

[2022-04-04T17:41:21.357] [DEBUG] default - Terraform v1.1.7
on darwin_amd64

Generated Terraform code for the stacks: stage
✨  Done in 8.33s.
```
</details>

### Expected Behavior

Outputs should escape all tokens:

```json
  "output": {
    "additional_output": {
      "value": "${random_string.id.result}"
    },
    "direct_output": {
      "value": "${random_string.id.result}"
    },
    "formatted_output": {
      "value": "${format(\"%s\", random_string.id.result)}"
    }
  },
```

### Actual Behavior

One token is not escaped (see line 3):

```json
  "output": {
    "additional_output": {
      "value": "random_string.id.result"
    },
    "direct_output": {
      "value": "${random_string.id.result}"
    },
    "formatted_output": {
      "value": "${format(\"%s\", random_string.id.result)}"
    }
  },
```

### Steps to Reproduce

Here's a snippet for typescript:

```typescript
// @ts-ignore
import { Fn, TerraformOutput, TerraformStack } from 'cdktf';
import { Construct } from 'constructs';
import { RandomProvider, StringResource } from '@cdktf/provider-random';

export class MyStack extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);

    new RandomProvider(this, 'random');

    const resource = new StringResource(this, 'id', { length: 1 });
    const nested = { value: resource.result };
    new TerraformOutput(this, 'direct_output', {
      value: resource.result,
    });
    new TerraformOutput(this, 'additional_output', {
      value: nested.value,
    });
    new TerraformOutput(this, 'formatted_output', {
      // if you replace the Fn.format call with a static string, everything works
      // value: 'static',
      value: Fn.format('%s', [nested.value]),
    });
  }
}
```

If it is run as above, the `additional_output` will not be escaped in the json output. Interestingly this happens only for the mapped/nested variable (i.e. the one that's wrapped in an object). It probably has something to do with the new token handling since `0.9.0`, because the exact same code works with `0.8.*`. I hope I didn't overlook something obvious when I attempted to upgrade our code to the newer version. Also, if the title doesn't fit, please feel free to adjust it.

### LINKED ISSUE #1698: Access network_interface generates an error
### cdktf & Language Versions

cdktf version: **0.10.1**
language: **python**

### Affected Resource(s)

Access instance details

```
  #Query a GCP Image filtered by Family        
        gcp_image = DataGoogleComputeImage(self, id = "windows_image", family = "windows-2019", project="windows-cloud" )

        #Get the network interface
        net_int = ComputeInstanceNetworkInterface(subnetwork ="<SUBNETWORK>",subnetwork_project = "<SHARED-PROJECT>")

        #Create a new vm
        instance = ComputeInstance(self,id = "sample-vm", machine_type="e2-standard-4", name="cdktf-instance-1", 
                        boot_disk= ComputeInstanceBootDisk(initialize_params=ComputeInstanceBootDiskInitializeParams(image=gcp_image.name)),
                        network_interface=[net_int], zone= "us-east4-a"
        )

        #Output 
        TerraformOutput(self, "instance_ip", instance.network_interface[0].network_ip)
```

### Debug Output

```
    Traceback (most recent call last):
      File "/Users/user/terraform-learning/cdktf-gcp/gcp/main.py", line 37, in <module>
        GcpBasicVmStack(app, "gcp")
      File "/Users/user/.local/share/virtualenvs/gcp-XHqyQSdy/lib/python3.9/site-packages/jsii/_runtime.py", line 86, in __call__
        inst = super().__call__(*args, **kwargs)
      File "/Users/user/terraform-learning/cdktf-gcp/gcp/main.py", line 34, in __init__
        TerraformOutput(self, "instance_ip", instance.network_interface[0].network_ip)
    TypeError: 'InterfaceDynamicProxy' object is not subscriptable
```

### Expected Behavior

Should be able to retrieve the instance ip in the output

### Actual Behavior

Exception

### References

- #0000

### PR DESCRIPTION
**Pull request by jsteinich**
Closes #1610 
Fixes #1447
Fixes #1371
Fixes #1363
Fixes #1520
Fixes #1684
Fixes #25
Fixes #1698
Fixes #434
Fixes #518
Fixes #663
Fixes #1087

See #1610 for more information.

This is technically a breaking change. Work has been done to reduce the likelihood of impacting users; however, return types for any list/set/map of complex properties is changing.

### CODE REVIEW

File: packages/cdktf/lib/complex-computed-list.ts [hunk 1]
```diff
@@ -354,6 +362,7 @@ abstract class MapList
     protected wrapsSet: boolean
   ) {
     super(terraformResource, terraformAttribute);
+    this.savedFqn = this.fqn;
```
Comment (DanielMSchmidt) [inline packages/cdktf/lib/complex-computed-list.ts:377]: Why is `savedFqn` needed?

File: packages/cdktf/lib/tfExpression.ts [hunk 2]
```diff
@@ -171,6 +172,11 @@ function markAsInner(arg: any) {
     return;
   }
 
+  if (arg instanceof Construct) {
```
Comment (DanielMSchmidt) [inline packages/cdktf/lib/tfExpression.ts]: I wonder if this fixes the infinite loop problem we encountered in some of the converted code snippets

File: packages/cdktf/lib/tokens/private/resolve.ts [hunk 3]
```diff
@@ -61,11 +62,16 @@ export function resolve(obj: any, options: IResolveOptions): any {
     const context: IResolveContext = {
       preparing: options.preparing,
       scope: options.scope,
+      suppressBraces: options.previousContext?.suppressBraces,
```
Comment (DanielMSchmidt) [inline packages/cdktf/lib/tokens/private/resolve.ts:65]: Is this reset somewhere in the context chain if you have two nested contexts? E.g.: for `Fn.tolist(bar.baz, \`var-${index}\`)` should resolve to `"${tolist(element(foo.bar.baz, "val-${locals.index}"))}"` Sorry, can't escape the backticks in the first example

File: packages/@cdktf/provider-generator/lib/get/generator/emitter/attributes-emitter.ts [hunk 4]
```diff
@@ -129,7 +129,11 @@ export class AttributesEmitter {
         return `new ${att.type.name}(this, "${att.terraformName}")`;
       }
     } else {
-      return `new ${att.type.name}OutputReference(this, "${att.terraformName}")`;
+      if (att.type.name.includes("IResolvable")) {
```
[the remark this answers is not in the document]
Comment (jsteinich) [reply packages/@cdktf/provider-generator/lib/get/generator/emitter/attributes-emitter.ts:132]: It's this case; https://github.com/hashicorp/terraform-cdk/blob/4201a5e99baa7aec0a39634c2d1ba2b62f6b4802/packages/%40cdktf/provider-generator/lib/get/generator/models/attribute-type-model.ts#L113

It happens when an attribute isn't computed or is optional.

File: packages/cdktf/lib/tokens/private/resolve.ts (hunk 3 above)
Comment (jsteinich) [reply packages/cdktf/lib/tokens/private/resolve.ts:65]: It's not reset, but `TFExpression.resolveString()` has code to handle adding them based on the tokens that exist.

File: packages/@cdktf/provider-generator/lib/get/generator/emitter/attributes-emitter.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (DanielMSchmidt) [reply packages/@cdktf/provider-generator/lib/get/generator/emitter/attributes-emitter.ts:132]: 👍

File: packages/cdktf/lib/tokens/private/resolve.ts (hunk 3 above)
Comment (DanielMSchmidt) [reply packages/cdktf/lib/tokens/private/resolve.ts:65]: Ah nice 👍

**Comment by ansgarm** [comment]
I think we should write an upgrade guide ([similar as to for v0.10](https://www.terraform.io/cdktf/release/upgrade-guide-v0-10)) after merging this.

As the PR is too large, I can't really open the review tab, so I'll just post it here:
* The changelog should contain a note about the changed behaviour for all lists (it currently only mentions the maps). It might explain this in short (and later link to the upgrade guide, as soon as we add it).

**Pull request merged**